qiskit/dagcircuit/dagdependency.py

```diff
@@ -454,6 +454,7 @@ def _gather_succ(self, node_id, direct_succ):
             the lists of direct successors are put into a single one
         """
         gather = self._multi_graph
+        gather.get_node_data(node_id).successors = []
```

Comment:
I am guessing that `DagDependency` was originally created to be primarily used in `TemplateOptimization` pass. For instance, look at the fields of `DAGDepNode`:
```
__slots__ = ["type", "_op", "name", "_qargs", "cargs", "sort_key", "node_id", "successors", "predecessors", "reachable", matchedwith", "isblocked", "successorstovisit", "qindices", "cindices"]
```
many of these only make sense in the context of `TemplateOptimization`.

For caching vs. not caching lists of direct and transitive predecessors and successors, there is likely some tradeoff between performance and memory consumption, depending on how many times `TemplateOptimization` happens to call these. I haven't really experimented with `TemplateOptimization`, so I don't know, but for this PR I don't want to risk making `TemplateOptimization` even slower than it is now, even though I completely agree that it probably makes perfect sense to use `self._multi_graph.successors` instead of `self.get_node_data(node_id).successors`. I do want this PR, as an enabler for block collection strategies that take commutativity into account (#8319), as these are currently slow and extremely memory-hungry to be applied to large circuits.

I do think it would be a good idea to refactor `DagDepNode`, `DagDependency` and `TemplateOptimization` pass.

Reply:
I have added the "TODO" comment. Just to be clear, I think that we should definitely refactor `DAGDependency`, however a lot of it might go hand-in-hand with doing changes to `TempateOptimization` and I did not want to touch this in the current PR.